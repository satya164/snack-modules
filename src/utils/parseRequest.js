/* @flow */

import querystring from 'querystring';

type Result = {
  scope: ?string,
  id: ?string,
  tag: ?string,
  deep: ?string,
  platform: ?string,
};

export default function parseRequest(url: string): ?Result {
  const match = /^\/(?:@([^/?]+)\/)?([^@/?]+)(?:\/([^@]+?))?(?:@(.+?))?(?:\?(.+))?$/.exec(
    url,
  );

  if (!match) {
    return null;
  }

  const scope = match[1]; // matches scope in `@sccope/package`
  const id = match[2]; // matches id in `@user/id` or `package`
  const deep = match[3]; // matches deep path in `package/debounce` or `package@^3.4.0/debounce`
  const tag = match[4] || 'latest'; // matches version number in `package@^3.4.0`
  const qs = match[5]; // matches the query string

  const query = qs ? querystring.parse(qs) : null;

  return {
    scope,
    id,
    tag,
    deep,
    platform: query && query.platform,
  };
}
