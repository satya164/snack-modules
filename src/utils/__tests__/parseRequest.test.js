/* @flow */
/* eslint-env jest */

import parseRequest from '../parseRequest';

it('gets id from request', () => {
  const parsed = parseRequest('/lodash');
  expect(parsed).toMatchSnapshot();
});

it('gets user from request', () => {
  const parsed = parseRequest('/@scoped/lodash');
  expect(parsed).toMatchSnapshot();
});

it('gets tag from request', () => {
  const parsed = parseRequest('/lodash@14.5.2');
  expect(parsed).toMatchSnapshot();
});

it('gets deep path from request', () => {
  const parsed = parseRequest('/lodash/bounce/index');
  expect(parsed).toMatchSnapshot();
});

it('gets deep path and tag from request', () => {
  const parsed = parseRequest('/lodash/bounce/index@14.5.2');
  expect(parsed).toMatchSnapshot();
});

it('gets tag from request', () => {
  const parsed = parseRequest('/lodash/bounce/index?platform=ios');
  expect(parsed).toMatchSnapshot();
});

it('gets all fields from request', () => {
  const parsed = parseRequest(
    '/@scoped/lodash/bounce/index@14.5.2?platform=ios',
  );
  expect(parsed).toMatchSnapshot();
});
