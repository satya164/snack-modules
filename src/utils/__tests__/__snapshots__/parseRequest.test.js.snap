// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`gets all fields from request 1`] = `
Object {
  "deep": "bounce/index",
  "id": "lodash",
  "platform": "ios",
  "scope": "scoped",
  "tag": "14.5.2",
}
`;

exports[`gets deep path and tag from request 1`] = `
Object {
  "deep": "bounce/index",
  "id": "lodash",
  "platform": null,
  "scope": undefined,
  "tag": "14.5.2",
}
`;

exports[`gets deep path from request 1`] = `
Object {
  "deep": "bounce/index",
  "id": "lodash",
  "platform": null,
  "scope": undefined,
  "tag": "latest",
}
`;

exports[`gets id from request 1`] = `
Object {
  "deep": undefined,
  "id": "lodash",
  "platform": null,
  "scope": undefined,
  "tag": "latest",
}
`;

exports[`gets tag from request 1`] = `
Object {
  "deep": undefined,
  "id": "lodash",
  "platform": null,
  "scope": undefined,
  "tag": "14.5.2",
}
`;

exports[`gets tag from request 2`] = `
Object {
  "deep": "bounce/index",
  "id": "lodash",
  "platform": "ios",
  "scope": undefined,
  "tag": "latest",
}
`;

exports[`gets user from request 1`] = `
Object {
  "deep": undefined,
  "id": "lodash",
  "platform": null,
  "scope": "scoped",
  "tag": "latest",
}
`;
